onnxruntime/core/providers/xnnpack/nn/conv.cc

```diff
@@ -535,9 +554,150 @@ Status Conv::Compute(OpKernelContext* context) const {
+  is_packed = false;
+  // only layout of weight input is adjusted via PrePack
+  if ((conv_type_ == OpComputeType::op_compute_type_fp32 && input_idx == 1) ||
+      (conv_type_ != OpComputeType::op_compute_type_fp32 && input_idx == 3)) {  // InputTensors::IN_W
+    if (xnn_conv_attrs_.group > 1) {
+      // Xnnpack [G, Oc, H, W Ic/G]
+      // (ref: https://github.com/google/XNNPACK/blob/ecd8311c8fd3d9ab47edbc3df5f2b5de7dabe75f/test/deconvolution-operator-tester.h#L678)
+      TensorShape orig_shape =
+          {xnn_conv_attrs_.group,
+           tensor.Shape()[0] / xnn_conv_attrs_.group,
+           tensor.Shape()[1],
+           tensor.Shape()[2],
+           tensor.Shape()[3]};
+
+      InlinedVector<size_t> perm{0, 2, 3, 4, 1};
+      TensorShapeVector new_dims{
+          orig_shape[0],
+          orig_shape[2],
+          orig_shape[3],
+          orig_shape[4],
+          orig_shape[1]};
+
+      packed_w_ = Tensor(tensor.DataType(), TensorShape(new_dims), std::move(alloc));
+      // g I/g O H W --> g O H W I/g
+      SingleAxisTranspose(perm, tensor, packed_w_, /*from*/ 1, /*to*/ 4, &orig_shape);
+    } else {
+      // Transpose from {M, C/group, kH, kW} to {M, kH, kW, C/group}
+
+      auto orig_shape = tensor.Shape();
+      InlinedVector<size_t> perm{1, 2, 3, 0};
+      TensorShapeVector new_dims{orig_shape[1],
+                                 orig_shape[2],
+                                 orig_shape[3],
+                                 orig_shape[0]};
+
+      packed_w_ = Tensor(tensor.DataType(), TensorShape(new_dims), std::move(alloc));
+      // I O H W --> O H W I
+      SingleAxisTranspose(perm, tensor, packed_w_, /*from*/ 0, /*to*/ 3);
+    }
+    is_packed = true;
+
+    // we can create the kernel now
+    auto ret = CreateXnnpackKernel(static_cast<ConvAttributes*>(&xnn_conv_attrs_), C_, M_, kernel_shape_, clip_min_max_, packed_w_,
+                                   B_, op0_,
+#ifdef XNN_CACHE_ENABLE
+                                   &xnn_caches_,
+#else
+                                   0,
+#endif
+                                   quant_param_, conv_type_, true);
+    ORT_RETURN_IF_ERROR(ret);
+  }
+
+  return Status::OK();
+}
+
+Status ConvTranspose::Compute(OpKernelContext* context) const {
+  const Tensor& X = *context->Input<Tensor>(0);  // this is in NHWC format
+  const auto& X_shape = X.Shape();
+  const int64_t N = X_shape[0];  // input is NHWC
+  const int64_t H = X_shape[1];
+  const int64_t W = X_shape[2];
+
+  TensorShapeVector Y_dims;
+  TensorShape input_shape = {H, W};
+
+  ConvAttributes::ConvPadVector pads(xnn_conv_attrs_.pads);
+  xnn_conv_attrs_.ComputePadsAndOutputShape(
+      input_shape, M_, kernel_shape_,
+      xnn_conv_attrs_.strides,
+      xnn_conv_attrs_.dilations, xnn_conv_attrs_.output_padding,
+      N, &pads, &Y_dims);
+
+  // convert to nhwc
+  Y_dims[1] = Y_dims[2];
+  Y_dims[2] = Y_dims[3];
+  Y_dims[3] = M_;
+  Tensor* Y = context->Output(0, TensorShape(Y_dims));
```

Comment:
`xnn_conv_attrs_.ComputePadsAndOutputShape` takes in  {H, W}, So it will never know what's the tensor layout. So it has a hypothesis all tensors are NCHW.